core/trino-main/src/main/java/io/trino/sql/planner/optimizations/AbstractMaterializeTableScan.java

```diff
@@ -0,0 +1,221 @@
+    }
+
+    protected abstract Result apply(T node, Captures captures, Session session);
+
+    protected Optional<List<Expression>> materializeTable(Session session, TableScanNode tableScan, Constraint constraint)
+    {
+        TableHandle table = tableScan.getTable();
+        List<ColumnHandle> columns = tableScan.getOutputSymbols().stream()
+                .map(symbol -> tableScan.getAssignments().get(symbol))
+                .toList();
+        List<Type> types = tableScan.getOutputSymbols().stream()
+                .map(Symbol::type)
+                .toList();
+
+        Future<Optional<List<Expression>>> future = executor.submit(() -> {
+            try (var _ = new SetThreadName("MaterializeTableScan-%s", table.catalogHandle())) {
+                return doMaterializeTable(session, table, columns, types, constraint);
+            }
+        });
+
+        Duration timeout = getMaterializeTableTimeout(session);
+        try {
+            return future.get(timeout.toMillis(), MILLISECONDS);
+        }
+        catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            future.cancel(true);
+            throw new RuntimeException("interrupted", e);
+        }
+        catch (ExecutionException e) {
+            if (e.getCause() instanceof TableTooLargeException) {
+                log.warn("Table too large to materialize: %s", table);
+            }
+            else {
+                log.warn(e.getCause(), "Failed to materialize table: %s", table);
+            }
+        }
+        catch (TimeoutException e) {
+            log.warn("Timed out while materializing table: %s", table);
+            future.cancel(true);
+        }
+        return Optional.empty();
+    }
+
+    private Optional<List<Expression>> doMaterializeTable(Session session, TableHandle table, List<ColumnHandle> columns, List<Type> types, Constraint constraint)
+            throws IOException
+    {
+        SplitSource splitSource = splitManager.getSplits(session, session.getQuerySpan(), table, DynamicFilter.EMPTY, constraint);
+        List<Split> splits = new ArrayList<>();
+        while (!splitSource.isFinished()) {
+            if (Thread.currentThread().isInterrupted()) {
+                throw new InterruptedIOException();
+            }
+            List<Split> batch = getFutureValue(splitSource.getNextBatch(1000)).getSplits();
+            for (Split split : batch) {
+                if (!split.isRemotelyAccessible() && !split.getAddresses().contains(currentNode)) {
+                    return Optional.empty();
+                }
+            }
+            splits.addAll(batch);
+            if (splits.size() > MAX_SPLITS) {
+                return Optional.empty();
+            }
+        }
+
+        PageSourceProvider pageSourceProvider = pageSourceManager.createPageSourceProvider(table.catalogHandle());
+        int maxRows = getMaterializeTableMaxActualRowCount(session);
```

Comment:
JDBC queries can be huge while they only yield few results. DF can change JDBC query duration from hours to seconds

Reply:
Do we have a way to estimate that? We would want to skip those.

Reply:
I don't think we have a way to estimate the cost of the work done by the pushed down operations into table scan. 
Every pushdown into a JDBC table handle will result in `io.trino.sql.planner.plan.TableScanNode#statistics` being revised with the estimate of the resulting output and we lose the information about estimate of original table at that point. One possibility is to remember the original table scan output estimate in a separate field and rely on that.
We had a somewhat similar need for avoiding a particular optimization on JDBC connectors in https://github.com/trinodb/trino/pull/22355 and ended up adding a new API `io.trino.spi.connector.ConnectorMetadata#allowSplittingReadIntoMultipleSubQueries` for that. Maybe we can explore generalizing/re-using that for this case too.
cc: @martint